Thank you working on this. This will allow shared module be much useful feature for broader set of projects. Here are my 5 cents:

## Errors

When shared feature is missing we should clearly explain users what happened. E.g. following flow: 

```cython
from array import array

def add_one(int[:] buf):
    for x in range(buf.shape[0]):
        buf[x] += 1


a = array('i', [1, 2, 3, 4])
add_one(a)
print(a.tolist())
```

If user by mistake exclude the memoryview from shared utility generation code:
```
cython --generate-shared shared.c  --shared-exclude MemoryView
cython --shared shared memview.pyx
```

Compilation will succeed but importing module will cause cryptic error:
```
>>> import memview
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    import memview
  File "memview.pyx", line 1, in init memview
    from array import array
ImportError: shared does not export expected C function __Pyx_ValidateAndInit_memviewslice
```

Should print error like "Add feature MemoryView to shared utility." Or something like that. **Ideally** we should error out during compilation but not sure if there is robust way to do it.

## CLI

* We need to clarify better when `--shared-exlude` and `--shared-only` needs to be used. If I am not wrong they should be used together with `--generate-shared` argument. They should error out otherwise.

* Related to new arguments, let me bikeshed a little. Have you consider use `feature` e.g. `--feature-exclude`, `--feature-only`.
I have not strong opinion about that. Maybe it fits better for CLI proposed lower (and hence not for this PR) 

* Long term (not this PR of course) I would do overhaul of whole CLI because it's getting complicated. We can:
  - `cython MODULE.pyx` -> `cython compile MODULE.pyx`
  - `cython --shared cyshared MODULE.pyx` -> `cython compile --shared MODULE.pyx`
  - `cython --generate-shared cyshared.c` -> `cython generate-shared cyshared.c`
  - `cython --generage-shared cyshared.c --shared-exclude MemoryView` -> `cython generate-shared --shared-exclude MemoryView cyshared.c`
  - `cythonize -fib MODULE.pyx` -> `cython build -fi MODULE.PYX`

## setuptools

We should have a simple and stupid way how to integrate it to setuptools:

```python
from Cython.Build import cythonize
from Cython.Compiler import Options
from setuptools import setup, Extension

extensions = [
    Extension("*", ["**/*.pyx"]),
    # Providing 'sources' is optional for the shared module.
    # If missing, the module package will be used for the path in 'build_dir'.
    Extension("mypkg.shared._cyutility", sources=["mypkg/shared/_cyutility.c"], shared_exclude='MemoryView'),
]

setup(
  ext_modules = cythonize(extensions, shared_utility_qualified_name = 'mypkg.shared._cyutility')
)
```

--- a/Cython/Build/SharedModule.py
+++ b/Cython/Build/SharedModule.py
@@ -54,9 +54,10 @@ def _iter_exports(selected_features=None):
         current_feature_name = file_base_name.partition('_')[0]
 
         if selected_features is not None and current_feature_name not in selected_features:
+            # Disable whole file feature, including any @features declared therein.
             continue
 
-        next_match = []  # Look-ahead to allow parsing "//@tags".
+        current_section = None
         for line in Code.read_utilities_hook(c_utility_file):
             if '//' not in line or not line.startswith('//'):
                 continue
@@ -66,24 +67,29 @@ def _iter_exports(selected_features=None):
 
             name = match.group('name')
             if name:
-                if '.export' not in name:
-                    continue
-                section_title = UtilityCode.match_section_title(name)
-                if not section_title:
-                    continue
-                name, section_type = section_title.groups()
-                if section_type == 'export':
-                    if next_match:
-                        yield next_match
-                    next_match = [current_feature_name, name, c_utility_file]
+                if current_section:
+                    if selected_features is None or current_section[0] in selected_features:
+                        yield current_section
+                current_section = None
 
-            elif (tag := match.group('tag')) and tag.startswith('feature') and next_match:
-                # Overwrite the file based feature name with the explicit @feature name.
+                section_title = UtilityCode.match_section_title(name) if '.export' in name else None
+                if section_title:
+                    name, section_type = section_title.groups()
+                    if section_type == 'export':
+                        current_section = [current_feature_name, name, c_utility_file]
+
+            elif (tag := match.group('tag')) and tag.startswith('feature') and current_section:
+                if current_section[0] != current_feature_name:
+                    # Ignore file feature name if at least one tag was declared.
+                    if selected_features is None or current_section[0] in selected_features:
+                        yield current_section
+
                 explicit_feature_name = tag.split(':', 1)[-1].strip()
-                next_match[0] = explicit_feature_name
+                current_section[0] = explicit_feature_name
 
-        if next_match:
-            yield next_match
+        if current_section:
+            if selected_features is None or current_section[0] in selected_features:
+                yield current_section
 
 
 def create_shared_library_pipeline(context, scope, options, result, selected_features=None):

--- a/Cython/Compiler/CmdLine.py
+++ b/Cython/Compiler/CmdLine.py
@@ -289,16 +289,24 @@ def parse_command_line(args):
 
     if options.use_listing_file and len(sources) > 1:
         parser.error("cython: Only one source file allowed when using -o\n")
+
     if options.shared_c_file_path:
         if len(sources) > 0:
             parser.error("cython: Source file not allowed when using --generate-shared\n")
     elif len(sources) == 0 and not options.show_version:
         parser.error("cython: Need at least one source file\n")
+
+    if options.shared_utility_features_disabled or options.shared_utility_features_enabled:
+        if not options.shared_c_file_path:
+            parser.error("cython: --shared-only and --shared-exclude require --generate-shared")
+
     if Options.embed and len(sources) > 1:
         parser.error("cython: Only one source file allowed when using --embed\n")
+
     if options.module_name:
         if options.timestamps:
             parser.error("cython: Cannot use --module-name with --timestamps\n")
         if len(sources) > 1:
             parser.error("cython: Only one source file allowed when using --module-name\n")
+
     return options, sources

--- a/tests/run/shared_utility_module.srctree
+++ b/tests/run/shared_utility_module.srctree
@@ -16,6 +16,10 @@ PYTHON -c "import test_shared_utility"
 PYTHON delete_generated_files.py
 PYTHON setup_memoryview_without_extensions.py build_ext --inplace --force -j3
 PYTHON -c "import pkg1.pkg11.feature_memoryview as mv; x = mv.test([1,2,3]); assert x == 6, x"
+
+PYTHON delete_generated_files.py
+PYTHON setup_extensions_without_autopickle.py build_ext --inplace --force -j3
+PYTHON -c "import pkg1.pkg11.feature_extensiontypes as ext; assert hasattr(ext.ExtensionType(), 'x')"
 """
 
 
@@ -61,6 +65,8 @@ setup(
 
 import pathlib
 assert pathlib.Path("pkg2/CythonShared.c").exists()
+with open("pkg2/CythonShared.c") as f:
+    assert '__Pyx_setup_reduce' not in f.read()  # AutoPickle indirectly disabled
 
 
 ######## setup_with_builddir.py ########
@@ -86,6 +92,9 @@ import pathlib
 assert pathlib.Path("build/pkg2/CythonShared.c").exists()
 assert not pathlib.Path("pkg2/CythonShared.c").exists()
 
+with open("build/pkg2/CythonShared.c") as f:
+    assert '__Pyx_setup_reduce' not in f.read()  # AutoPickle indirectly disabled
+
 
 ######## setup_without_sources.py ########
 
@@ -109,6 +118,9 @@ import pathlib
 assert pathlib.Path("pkg2/CythonShared.c").exists()
 assert not pathlib.Path("build/pkg2/CythonShared.c").exists()
 
+with open("pkg2/CythonShared.c") as f:
+    assert '__Pyx_setup_reduce' not in f.read()  # AutoPickle indirectly disabled
+
 
 ######## setup_memoryview_without_extensions.py ########
 # Memoryviews depend on (some) extension type support themselves.
@@ -128,6 +140,38 @@ setup(
         shared_utility_features_disabled = ['ExtensionTypes'],
     )
 )
+
+import pathlib
+assert pathlib.Path("pkg2/CythonShared.c").exists()
+
+with open("pkg2/CythonShared.c") as f:
+    assert '__Pyx_setup_reduce' in f.read()  # MemoryViews use AutoPickle
+
+
+######## setup_extensions_without_autopickle.py ########
+
+from Cython.Build import cythonize
+from setuptools import setup, Extension
+
+extensions = [
+    Extension("*", ["pkg1/pkg11/feature*.pyx"]),
+    Extension("pkg2.CythonShared", sources=["pkg2/CythonShared.c"]),
+]
+
+setup(
+    ext_modules = cythonize(
+        extensions,
+        shared_utility_qualified_name = 'pkg2.CythonShared',
+        shared_utility_features_disabled = ['AutoPickle', 'MemoryView'],
+        compiler_directives={'auto_pickle': False},
+    )
+)
+
+import pathlib
+assert pathlib.Path("pkg2/CythonShared.c").exists()
+
+with open("pkg2/CythonShared.c") as f:
+    assert '__Pyx_setup_reduce' not in f.read()
 
 
 ######## test_shared_utility.py ########
@@ -201,7 +245,7 @@ def imported_func():
 # Using feature "ExtensionTypes".
 
 cdef class ExtensionType:
-    cdef int x
+    cdef public int x
 
 
 ######## pkg1/pkg11/feature_memoryview.pyx ########

--- a/tests/run/shared_utility_module.srctree
+++ b/tests/run/shared_utility_module.srctree
@@ -12,6 +12,10 @@ PYTHON -c "import test_shared_utility"
 PYTHON delete_generated_files.py
 PYTHON setup_without_sources.py build_ext --inplace --force -j3
 PYTHON -c "import test_shared_utility"
+
+PYTHON delete_generated_files.py
+PYTHON setup_memoryview_without_extensions.py build_ext --inplace --force -j3
+PYTHON -c "import pkg1.pkg11.feature_memoryview as mv; x = mv.test([1,2,3]); assert x == 6, x"
 """
 
 
@@ -106,6 +110,26 @@ assert pathlib.Path("pkg2/CythonShared.c").exists()
 assert not pathlib.Path("build/pkg2/CythonShared.c").exists()
 
 
+######## setup_memoryview_without_extensions.py ########
+# Memoryviews depend on (some) extension type support themselves.
+
+from Cython.Build import cythonize
+from setuptools import setup, Extension
+
+extensions = [
+    Extension("*", ["pkg1/pkg11/feature*.pyx"]),
+    Extension("pkg2.CythonShared", sources=["pkg2/CythonShared.c"]),
+]
+
+setup(
+    ext_modules = cythonize(
+        extensions,
+        shared_utility_qualified_name = 'pkg2.CythonShared',
+        shared_utility_features_disabled = ['ExtensionTypes'],
+    )
+)
+
+
 ######## test_shared_utility.py ########
 
 from pkg1.pkg11.add_one import add_one, imported_func
@@ -124,14 +148,14 @@ s = fstring()
 assert s == ' 5 != 10 ', s
 
 try:
-    from pkg1.pkg11.missing_feature_ExtensionTypes import ExtensionType
+    from pkg1.pkg11.feature_extensiontypes import ExtensionType
 except ImportError:
     pass
 else:
     assert False, "Import with missing feature 'ExtensionTypes' succeeded !"
 
 try:
-    from pkg1.pkg11.missing_feature_MemoryView import memview_func
+    from pkg1.pkg11.feature_memoryview import memview_func
 except ImportError:
     pass
 else:
@@ -173,18 +197,23 @@ def imported_func():
     return 1
 
 
-######## pkg1/pkg11/missing_feature_ExtensionTypes.pyx ########
-# Using a disabled feature.
+######## pkg1/pkg11/feature_extensiontypes.pyx ########
+# Using feature "ExtensionTypes".
 
 cdef class ExtensionType:
     cdef int x
 
 
-######## pkg1/pkg11/missing_feature_MemoryViews.pyx ########
-# Using a disabled feature.
+######## pkg1/pkg11/feature_memoryview.pyx ########
+# Using feature "MemoryViews".
+
+def test(ints):
+    from array import array
+    a = array('i', ints)
+    return memview_func(a)
 
 def memview_func(int[:] ints):
-    return sum(i for i in range(len(ints)))
+    return sum(ints[i] for i in range(len(ints)))
 
 
 ######## pkg2/__init__.py ########

